### ISSUE TITLE
Consider ways of differentiating implementation and poor error pages

### ISSUE THREAD

**Issue by oliverdunk** [WICG/change-password-url#14]
I've been experimenting with the `.well-known/change-password` route on a number of sites. On some, such as `https://twitch.tv/`, it's currently impossible to detect programatically that the route does not exist.

There are two reasons why this problem occurs:

- You see a redirect, as Twitch redirect users from their "naked" domain to the `www` subdomain.
- Twitch is a SPA, and doesn't seem to set status codes consistently. You see a 200 on the page you're redirected to, even though it's a 404.
_Requests occasionally go to the cache, and a 404 is correctly returned, so you may need to visit a number of times to observe this behaviour._

While in this example the server is setup incorrectly, for clients wishing to implement this spec, this is still a problem. It would be nice if servers were expected to exhibit additional behaviour (setting a header perhaps) so that more reliable detection could be done.

For my use case, I'm actually making CORS requests, so only know the status code and if we were redirected. It might not be possible for a solution to work for my use case - I certainly can't read headers - but am mentioning in case there is a CORS compatible option here.

**Comment by groovecoder** [comment]
I'm running into this same issue with many websites having sloppy and/or global redirect rules that have nothing to do with `.well-known/change-password` that make it messy to programmatically determine whether the redirect is the change-password page, or just a global redirect rule.

E.g.,
* https://adobe.com/.well-known/change-password goes to https://www.adobe.com/.well-known/change-password
* https://www.cafepress.com/.well-known/change-password goes to https://www.cafepress.com/+gifts
* https://www.elance.com/.well-known/change-password goes to https://www.upwork.com/.well-known/change-password
etc.

**Comment by rmondello** [comment]
There are a number of issues here. I’ll try to take them one by one, but not necessarily in order.

> https://adobe.com/.well-known/change-password goes to https://www.adobe.com/.well-known/change-password

To more reliably determine whether the user would ultimately be taken to a valid Change Password page, I recommend following any redirects that are returned. This is represented in this spec as:

> Clients must handle such redirects when requesting a change password url.

So, in this case, perform the same operation with the redirected-to URL, and notice that `https://www.adobe.com/.well-known/change-password` is a 404.

> https://www.elance.com/.well-known/change-password goes to https://www.upwork.com/.well-known/change-password

Then, perform the same operation with the redirected-to URL, and notice that `https://www.upwork.com/.well-known/change-password` is a 404.



> I've been experimenting with the .well-known/change-password route on a number of sites. On some, such as https://twitch.tv/, it's currently impossible to detect programatically that the route does not exist.
> https://www.cafepress.com/.well-known/change-password goes to https://www.cafepress.com/+gifts

For Twitch, just like the above cases, follow the redirect to eliminate the first problem.

That leaves us with this problem: Some websites don’t set status codes correctly. More specifically, some websites send a 200 status code for a web page that’s really a Not Found page, a 404.

We’ve run into similar problems, and have been thinking about this. We plan on writing up a full proposal for mitigating this, but I’ll briefly describe it here to get y’all’s feedback! (Disclaimer: This is an early, rough sketch of a proposal.)

We can validate if a 200 status code from a server for a Well-Known URL (or its ultimate destination, following redirects) is trustworthy by requesting a URL we’d expect to never be a 200. To determine whether the URL otherwise successfully resolved by the Change Password URL is meaningful, also perform the following:
- Request `example.com/.well-known/resource-that-should-not-exist-whose-status-code-should-not-be-200`, or similar (e.g. `example.com/.well-known/a-resource-that-should-not-exist` or `example.com/.well-known/non-existing-resource`).
- If the eventual status code for that load, after following redirects, is 200, assume that the server is not configured properly, and determine that the website in question doesn’t properly support the Change Password URL.

To summarize:
1. Solving the first problem, as I outline, can be done by following redirects and looking at the ultimate destination of redirects.
2. Solving the second problem involves a slightly more involved technique that we suspect will work, requires some testing, and should be formalize somewhere.

**Comment by oliverdunk** [comment]
Hey @rmondello! Thank you for taking the time to write such a detailed response.

I really like your proposal. It's made especially sweet given that it works with CORS, and requires no changes by most developers.

Do you think it's important that this `.well-known` route is only for the change password spec to use? I can imagine a scenario down the line, where a developer implements a 404 on that route specifically, to get another feature that also checks it to work. We then run into the same ambiguity about if the change password route is implemented or not.

**Comment by rmondello** [comment]
Sorry it took so long, @oliverdunk!

This is an interesting proposed well known URL, because its existence is for user agents and other software, and not really for websites to implement. Ideally, no website developer would ever come across this concept. Implementing this specific endpoint as a 404 without fixing their more general problem would be unideal, and would defeat the mitigation.

I think it makes sense for this new URL to be a standalone concept that’s referenced by the change password spec, so that it can be used to solve other similar problems in the future.

**Comment by groovecoder** [comment]
Yes, I like the approach to request some should-be-404 urls to detect site that are mis-configured to return 200's on the change password url. I'll use that in my current situation, and follow redirects as needed.

**Comment by hober** [comment]
Hi @groovecoder @oliverdunk @rmondello!

In 0b98642 I've defined this mechanism in [a separate spec in this repository](https://wicg.github.io/change-password-url/response-code-reliability.html).

I'm going to close this issue; please file new issues if there are problems with either document.
